pkg/transport/queuestorage/file.go

```diff
@@ -1,164 +1,328 @@
-	Queue  queue.QueueBackend
-	Notify chan FileEvent
-
-	MaxFileSize int64
-	MaxRows     int64
-	MaxFileAge  time.Duration
-}
+func NewFileWriter(param FileWriterParam) (*FileWriter, error) {
+	if param.MaxFileSize == 0 {
+		param.MaxFileSize = MaxFileSize
+	}
+	if param.MaxRows == 0 {
+		param.MaxRows = MaxRows
+	}
+	if param.MaxFileAge == 0 {
+		param.MaxFileAge = MaxFileAge
+	}
 
-func NewFileWriter(param NewFileWriterParam) (*FileWriter, error) {
 	fw := &FileWriter{
 		key:         param.Key,
-		path:        param.Path,
-		store:       param.Store,
-		queue:       param.Queue,
-		notify:      param.Notify,
 		maxFileSize: param.MaxFileSize,
 		maxRows:     param.MaxRows,
+		maxFileAge:  param.MaxFileAge,
 
-		ticker: time.NewTicker(time.Second),
-
-		expiry: time.Now().Add(param.MaxFileAge),
-	}
-
-	if fw.notify == nil {
-		return nil, fmt.Errorf("notify channel cannot be nil")
+		queue:   param.Queue,
+		storage: param.Storage,
 	}
 
-	if err := fw.create(); err != nil {
+	fileName := filepath.Join(param.Dir, fw.key, ulid.Make().String()+".ndjson")
+	if err := fw.create(fileName); err != nil {
 		return nil, err
 	}
 
 	return fw, nil
 }
 
-func (f *FileWriter) create() error {
+// create creates a file and all directories in its path.
+// It sets a timer which triggers close after the maxFileAge elapses.
+func (f *FileWriter) create(fileName string) error {
+	f.path = fileName
 	err := os.MkdirAll(filepath.Dir(f.path), os.ModePerm)
 	if err != nil {
-		log.Err(err).Msg("Unable to create all directories in file path")
+		log.Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to create all directories in file path")
 		return err
 	}
 
 	f.fd, err = os.OpenFile(f.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
 	if err != nil {
-		log.Err(err).Send()
+		log.Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Send()
 		return err
 	}
-	_ = f.fd.SetDeadline(f.expiry)
-
-	return nil
-}
 
-func (f *FileWriter) countDown() {
-	defer f.wg.Done()
-	defer f.ticker.Stop()
-
-	for {
-		select {
-		case now := <-f.ticker.C:
-			if now.After(f.expiry) {
-				log.Info().Str("key", f.key).
+	f.timer = time.AfterFunc(f.maxFileAge, func() {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+		if !f.terminated {
+			log.Info().
+				Str("key", f.key).
+				Str("filePath", f.path).
+				Msg("maximum file age reached, closing file")
+
+			if err := f.close(false); err != nil {
+				log.Err(err).
+					Str("key", f.key).
 					Str("filePath", f.path).
-					Msg("Maximum file age reached, closing file")
-				err := f.Close()
-				if err != nil {
-					log.Err(err).Send()
-				}
-				return
+					Msg("ticker unable to close file")
 			}
 		}
-	}
+	})
+
+	return nil
 }
 
-func (f *FileWriter) Write(data []byte) (n int, err error) {
-	f.canWrite.Lock()
-	defer f.canWrite.Unlock()
+// ensureWritable checks for constraints before any write operation.
+func (f *FileWriter) ensureWritable(dataSize int64) (err error) {
+	var openNew bool
```

Comment:
It's  not clear what `openNew` signifies, so it'd help to add a comment.

But I'd remove it and just `return f.close()`. Naked returns make it hard to see what is being returned and can also lead to bugs if variables are shadowed.